docs/madr/2-storing-registry-state.md

```diff
@@ -0,0 +1,107 @@
+which exposes the lua-sqlite3 library for using in-memory databases.
+
+Directly using Lua tables for complex data has shown a need for better tooling
+in manipulating and searching data, making the SQLite implementation a potential
+solution. However, its novelty raises concerns about reliability in larger
+products.
+
+## Decision Drivers
+
+The main drivers for this decision are:
+
+- **Scalability**
+  - The storage mechanism should be computation and storage efficient for
+    optimal resource usage and data fetching.
+- **Performance**
+  - Fast queries are desired.
+- **Ecosystem Integration**
+  - Exploring new modules and their potential benefits within the ao ecosystem.
+- **Maintenance**
+  - How many development cycles will be needed to maintain the solution.
+
+## Considered Options
+
+### Option 1: Using Lua Primative Tables
+
+- **Pros:**
+  - Known and stable.
+  - No need for additional testing or integration.
+  - zero overhead.
+- **Cons:**
+  - Demonstrated scalability issues when attempting to retrieve 150k records in
+    ANT's.
+  - Limited performance with larger datasets, namely deeply nested tables.
+  - Lack of strictness on data types.
+  - Lots of if-and-or-buts when interacting with them using certain combinations
+    of data types (for example lists are null terminated on some iterators)
+
+### Option 2: Use aos-sqlite Module
+
+- **Pros:**
+  - Addresses _compute_ scalability issues with direct table usage due to
+    efficient querying built into sqlite
+  - Potential for improved performance with larger and more complex datasets and
+    schemas
+  - Maintained by core team of AO.
+  - Additional flexibility in the ways we can store and query for data.
+- **Cons:**
+  - New and relatively untested in the aspect of range of applications of it in
+    the ecosystem.
+  - Higher maintenance of the implementation.
+  - additional overhead in setup
+  - lack of tooling to manage migrations
+  - lack of tooling around testing
+
+## Decision Outcome
+
+It was decided to use Lua tables for storing the state of the registry for the
+following reasons:
+
+- **In-Memory Limitation**: Given that in `aos`, everything is in-memory and we
+  can't write to disk, we lose out on certain features of SQLite.
+- **Overhead**: While SQLite provides some useful tooling, it introduces
+  significant overhead. A simple data structure with Lua table primitives can
+  implement the same state management with less complexity.
+- **Storage vs. Compute**: SQLite bloats the size of the data in an effort to
+  reduce the compute for accessing data. In `aos`, being in-memory, storage is
+  at a premium, so computing values is preferred over storing them (to an
+  extent).
+
+### Positive Consequences
+
+- **Scalability Improvements**: Direct use of Lua tables addresses storage
+  scalability issues.
+- **Resource Efficiency**: More efficient use of in-memory storage by computing
```

Comment:
What do you mean by computing values rather than storing them in this context?

Reply:
I was going by memory when I wrote this thinking back to our huddle yesterday but I realize now we ARE actually storing them huh? Was part of the point of implementing hash lookups. Will remove.